src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/DPBasedJoinPlanOptimizer.java

```diff
@@ -82,8 +87,50 @@ else if ( plan_left.getRootOperator() instanceof PhysicalOpRequestWithTranslatio
 
     }
 
-    // This method returns all subsets (with the given size) of the given superset.
-    public static <T> List<List<T>> getSubSet( final List<T> superset, final int n ){
+    public abstract <T> List<Pair<List<T>, List<T>>> splitIntoSubSets( final List<T> superset );
+
+    /**
+     * In cases in which there is a union with requests under right input,
+     * this function turns the requests into xxAdd operators with the previous join arguments as subplans.
+     **/
+    protected PhysicalPlan createPlansWithUnaryOpForUnionPlan( final PhysicalPlan inputPlan, final PhysicalPlan unionPlan ) {
+        final int numberOfSubPlansUnderUnion = unionPlan.numberOfSubPlans();
+        final PhysicalPlan[] newUnionSubPlans = new PhysicalPlan[numberOfSubPlansUnderUnion];
+
+        for ( int i = 0; i < numberOfSubPlansUnderUnion; i++ ) {
+            final PhysicalPlan oldSubPlan = unionPlan.getSubPlan(i);
+            final PhysicalPlan newSubPlan;
+
+            final PhysicalOperator oldSubPlanRootOp = oldSubPlan.getRootOperator();
+            if ( oldSubPlanRootOp instanceof PhysicalOpRequest ) {
+                final PhysicalOpRequest<?,?> reqOp = (PhysicalOpRequest<?,?>) oldSubPlanRootOp;
+                final UnaryLogicalOp addOp = LogicalOpUtils.createLogicalAddOpFromPhysicalReqOp(reqOp);
+                newSubPlan = PhysicalPlanFactory.createPlan( addOp, inputPlan);
+            }
+            else if ( oldSubPlanRootOp instanceof PhysicalOpFilter
+                    && oldSubPlan.getSubPlan(0).getRootOperator() instanceof PhysicalOpRequest ) {
+                final PhysicalOpFilter filterOp = (PhysicalOpFilter) oldSubPlanRootOp;
+                final PhysicalOpRequest<?,?> reqOp = (PhysicalOpRequest<?,?>) oldSubPlan.getSubPlan(0).getRootOperator();
+
+                final UnaryLogicalOp addOp = LogicalOpUtils.createLogicalAddOpFromPhysicalReqOp(reqOp);
+                final PhysicalPlan addOpPlan = PhysicalPlanFactory.createPlan( addOp, inputPlan);
+
+                newSubPlan = PhysicalPlanFactory.createPlan( filterOp, addOpPlan);
+            }
+            else {
+                newSubPlan = PhysicalPlanFactory.createPlanWithJoin( inputPlan, oldSubPlan);
+            }
```

Comment:
That's not right. We shouldn't get into that situation. I would say that this whole rewriting as done by this `createPlansWithUnaryOpForUnionPlan` function should be done only for cases in which the subplans under the union are all requests or filters with requests. Hence, before calling this function, test whether the given `unionPlan` is such a case.